keps/sig-apps/4212-improve-node-maintenance/README.md

```diff
@@ -0,0 +1,1065 @@
+  politely ask them to migrate their pods by creation of NodeMaintenance object. 
+- The PDB controller should detect and account for applications with max surge capability when
+  calculating PDB status.
+- Introduce a node maintenance controller that will help controllers like deployment controller
+  to migrate their pods.
+
+### Non-Goals
+- Introduce SurgeInterface, which CRDs could implement so that their pods would not be targeted
+  by node drain. Conceptually similar to ScaleInterface (https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/scale/interfaces.go).
+- Introduce a field that could include non-critical daemon set pods
+  (priority `system-cluster-critical` or `system-node-critical`) for node maintenance/drain request.
+  The daemon set controller would then gracefully shut down these pods. Critical pods could be
+  overridden by the priority list mentioned below.
+- NodeMaintenance could include a plan of which pods to target first. Similar to graceful node
+  shutdown, we could include a list of priorities to decide which pods should be terminated first.
+  This list could optionally include pod timeouts, but could also wait for all the pods of a given
+  priority class to finish first without a timeout. This could also be used to target daemon set
+  pods of certain priorities (see point above). We could also introduce drain profiles based on
+  these lists. The cluster admin could then choose or create such a profile based on his/her needs.
+  The logic for processing the decision list would be contained in the node maintenance controller,
+  which would set an intent to selected pods to shut down via the TargetedByNodeMaintenance
+  condition.
+- Introduce a node maintenance period or a TTL optional field as an upper bound on the duration of
+  node maintenance. Then the node maintenance would be garbage collected and the node made
+  schedulable again.
+
+## Proposal
+Most of these issues stem from missing a standardized way of detecting a start of the node drain.
+This KEP proposes the introduction of a NodeMaintenance object that would signal an intent to
+gracefully remove pods from a given node. This should complement the node's `.spec.unschedulable`,
+which prevents new pods from being scheduled on such a node. We will focus primarily on
+`kubectl drain`, but the API can also be used by other drain implementations as well (e.g. node
+autoscalers).
+
+We will first introduce the NodeMaintenance API and then later modify the behavior of the
+Kubernetes system to fix all the node drain issues we mentioned earlier. The name, NodeMaintenance
+is subject to change; please see the alternatives section for other suggestions.
+
+To support workload migration, a new controller should be introduced to observe the NodeMaintenance
+objects and then mark pods for migration with a condition. The pods would be selected according to
```

Comment:
This is meant as the application migration. I have tried to edit the places with migration to be more specific. Please let me know if you feel it is still confusing.

Pod restart could be also confused with the container restart, and technically during the workload move it will be a new instance of the pod.